File: 377-combination-sum-iv/377-combination-sum-iv.cpp
class Solution {
public:
    int combinationSum4(vector<int>& nums, int tar) {
        vector<int> dp(tar+1 , -1) ;
        return dfs(nums , tar , dp) ; 
    }
    

    int dfs(vector<int>& nums , int tar , vector<int>& dp){

        if(tar ==0) return 1 ; 
        if( dp[tar] != -1) return dp[tar] ; 

        dp[tar] = 0 ; 

        for(auto n:nums){
            if( tar- n >=0){
                dp[tar] +=  dfs(nums, tar-n , dp); 
            }}
        return dp[tar ] ; 
    }


    
};

File: 377-combination-sum-iv/377-combination-sum-iv.py
class Solution:
    def combinationSum4(self, nums: List[int], tar: int) -> int:
        
        
        @lru_cache(maxsize = None )
        def dfs(rem):
            
            if rem ==0: return 1 
            
            res = 0 
            for n in nums: 
                if rem - n >=0: 
                    res += dfs(rem-n)
            
            return res 
        
        return dfs(tar)